### DISCUSSION TITLE
Replicating XY(Z) resolution in different renderers when taking multiple screenshots

### DISCUSSION THREAD

**Discussion by supersubscript** [pyvista/pyvista#4807] [Q&A]
## Description
I have a number of meshes from flower primordia (early flower organs) that I would like to plot with the same spatial resolution for comparison. The problem is that I have quite a number of these, namely ~500, with each mesh containing up to 15k points. Because of this, it is way too memory intensive to plot everything in the same renderer. I want each individual image to be directly comparable to every other one.

How would I go about ensuring that the XY(Z) resolution is the same in two images when taking a screenshot? 

## Example Data
All primordia (arbitrary scales; to understand what I'm trying to visualize, but with correct scaling):
https://drive.google.com/open?id=1l7NFP1ABzeoLAd2rbzgvGAulfi2dlzWm
Note that the image is really big, so you might have to try different viewers to open it.

Two example primordia (plotted with different scales due to the perspective):
![Col0-24h-light-LowN-20_segmented_segmented_domain_6](https://user-images.githubusercontent.com/14143529/60022201-41331c80-968b-11e9-9730-07a1cc75b78b.png)
https://drive.google.com/file/d/1gVbgNPOGjeeGgJScmmZJuN3Vz8b4tzTD/view?usp=sharing

![Col0-24h-light-LowN-20_segmented_segmented_domain_3](https://user-images.githubusercontent.com/14143529/60022206-42fce000-968b-11e9-9278-3c24e749bc8a.png)
https://drive.google.com/file/d/1mD89FAM8C8EF4dG1tzp5sqXGtdcIiTs4/view?usp=sharing
```python
import pyvista as pv
import os

fnames = ['/home/henrik/Col0-24h-light-LowN-20_segmented_segmented_domain_6.ply', 
'/home/henrik/Col0-24h-light-LowN-20_segmented_segmented_domain_3.ply']

for fname in fnames:
    domain = pv.read(fname)
    p = pv.Plotter()
    p.add_mesh(domain, color='white')
    p.view_yz()
    p.set_background('white')
    p.screenshot('/home/henrik/' + os.path.basename(fname[:-4]) + '_domain_' + str(label) + '.png')
    p.close()
``` 

Note that the two primordia above look like they are similar in size due to how the plotting works, when the first mesh is actually much smaller than the second one.

**Comment by banesullivan** [comment]
Here is one way to do this by translating all datasets to be centered at `(0,0,0)` (via `domain.points -= domain.center`) and using a predefined camera position that focuses on the origin. This is equivalent to actually setting up a camera on a desk and focusing it at a point then putting a physical object at that point - and just like a real camera, if you're zoomed in too much, you could miss some of the physical object. To mitigate that, I go ahead and load all the datasets upfront into a `MultiBlock` dataset and find the largest object via the `length` attribute. Using that, I get the bounding box of that largest object and define a camera position for all scenes.

Note that if you don't want to load all of the meshes and store them in a `MultiBlock` dataset, then you could read, check the size and store the bounds until you have the largest one.

```py
import pyvista as pv
import os

fnames = ['Col0-24h-light-LowN-20_segmented_segmented_domain_6.ply', 
          'Col0-24h-light-LowN-20_segmented_segmented_domain_3.ply',]

domains = pv.MultiBlock()
max_length, idx = 0.0, 0
for i, fname in enumerate(fnames):
    domain = pv.read(fname)
    # center the domain at the origin
    domain.points -= domain.center
    domains[os.path.basename(fname[:-4])] = domain
    if domain.length > max_length:
        max_length = domain.length
        idx = i
# Now get a box of the max size
max_extent = domains[idx].outline()
```

```py
# Manually choose a camera position
cpos = max_extent.plot(cpos='yz', background='white', color='red', 
                       notebook=False, use_panel=False)
```

![download](https://user-images.githubusercontent.com/22067021/60037112-079fe880-966e-11e9-82f7-d16658f7c65c.png)


```py
# You don't need to show the box, but I did
show_box = True

for fname in fnames:
    domain = pv.read(fname)
    # center the domain at the origin
    domain.points -= domain.center
    
    p = pv.Plotter()
    p.add_mesh(domain, color='white')
    if show_box:
        p.add_mesh(max_extent, color='red')
    # Use predefined camera location so spatial scale is constant
    p.camera_position = cpos
    p.set_background('white')
    label = 'foo'
    p.screenshot(os.path.basename(fname[:-4]) + '_domain_' + str(label) + '.png')
    p.close()
```

![Col0-24h-light-LowN-20_segmented_segmented_domain_3_domain_foo](https://user-images.githubusercontent.com/22067021/60037234-42a21c00-966e-11e9-9dcc-610a9742dfdf.png)
![Col0-24h-light-LowN-20_segmented_segmented_domain_6_domain_foo](https://user-images.githubusercontent.com/22067021/60037235-42a21c00-966e-11e9-9cd7-87cdaf3af3f5.png)


And with no box:

![Col0-24h-light-LowN-20_segmented_segmented_domain_3_domain_foo](https://user-images.githubusercontent.com/22067021/60037297-6cf3d980-966e-11e9-8eb9-deeb5d0453ff.png)
![Col0-24h-light-LowN-20_segmented_segmented_domain_6_domain_foo](https://user-images.githubusercontent.com/22067021/60037298-6cf3d980-966e-11e9-8a4e-8378cd4e6609.png)

**Comment by banesullivan** [comment]
If you don't want the exact same camera position for all scenes, but want to keep the spatial resolution constant, you could take the camera position for one scene - and use its position to get a radius then use that radius to make a `Sphere`. Then you could move the camera around the nodes of that sphere and focus at the origin and the spatial resolution would be the same regardless of where you are on the sphere.

**Comment by banesullivan** [comment]
@supersubscript - did this solution work for you? Can we close this?

**Comment by supersubscript** [comment]
Sorry for the late reply, @banesullivan! Solution is working great -- thanks for the help!
